False positive in clang-analyzer-cplusplus.NewDeleteLeaks for std::unique_ptr

```
#include <memory>
#include <cstring>
#include <string>

struct Repro {
    std::unique_ptr<int> m_ptr;

    struct Other {
        char url[256];
    };
    char m_buff[1000];

    void CopyIn()
    {
        auto other = (Other *)m_buff;

        std::string local_url = "foo bar";
        strncpy(other->url, local_url.c_str(), sizeof(other->url) - 1);
    }
};

void
bug()
{
    Repro rep;
    rep.m_ptr = std::make_unique<int>();
    rep.CopyIn();
}
```

`clang-tidy` reports the following when building against the GCC-15's libstdc++:

```
repro.cpp:19:5: error: Potential leak of memory pointed to by field '_M_head_impl' [clang-analyzer-cplusplus.NewDeleteLeaks,-warnings-as-errors]
   19 |     }
      |     ^
repro.cpp:26:17: note: Calling 'make_unique<int, >'
   26 |     rep.m_ptr = std::make_unique<int>();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
/opt/gcc-15/lib/gcc/x86_64-pc-linux-gnu/15.1.1/../../../../include/c++/15.1.1/bits/unique_ptr.h:1085:30: note: Memory is allocated
 1085 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
repro.cpp:26:17: note: Returned allocated memory
   26 |     rep.m_ptr = std::make_unique<int>();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
repro.cpp:27:5: note: Calling 'Repro::CopyIn'
   27 |     rep.CopyIn();
      |     ^~~~~~~~~~~~
repro.cpp:19:5: note: Potential leak of memory pointed to by field '_M_head_impl'
   19 |     }
      |     ^
```

There is clearly no leak here and the warning goes away if the code is tweaked just a little bit... I cannot understand

--- a/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
@@ -2234,16 +2234,6 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
     // situation.
     bool CouldAccessOutOfBound = true;
     if (IsBounded && amountCopied.isUnknown()) {
-      // Get the max number of characters to copy.
-      const Expr *LenExpr = Call.getArgExpr(2);
-      SVal LenVal = state->getSVal(LenExpr, LCtx);
-
-      // Protect against misdeclared strncpy().
-      LenVal =
-          svalBuilder.evalCast(LenVal, sizeTy, LenExpr->getType());
-
-      std::optional<NonLoc> LenValNL = LenVal.getAs<NonLoc>();
-
       auto CouldAccessOutOfBoundForSVal = [&](NonLoc Val) -> bool {
         return !isFirstBufInBound(C, state, C.getSVal(Dst.Expression),
                                   Dst.Expression->getType(), Val,
@@ -2254,33 +2244,21 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
         CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*strLengthNL);
       }
 
-      if (CouldAccessOutOfBound && LenValNL) {
-        switch (appendK) {
-        case ConcatFnKind::none:
-        case ConcatFnKind::strcat: {
+      if (CouldAccessOutOfBound) {
+        // Get the max number of characters to copy.
+        const Expr *LenExpr = Call.getArgExpr(2);
+        SVal LenVal = state->getSVal(LenExpr, LCtx);
+
+        // Protect against misdeclared strncpy().
+        LenVal = svalBuilder.evalCast(LenVal, sizeTy, LenExpr->getType());
+
+        std::optional<NonLoc> LenValNL = LenVal.getAs<NonLoc>();
+
+        // Because analyzer doesn't handle expressions like `size -
+        // dstLen - 1` very well, we roughly use `size` for
+        // ConcatFnKind::strlcat here, same with other concat kinds.
+        if (LenValNL)
           CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*LenValNL);
-          break;
-        }
-        case ConcatFnKind::strlcat: {
-          if (!dstStrLengthNL)
-            break;
-
-          SVal FreeSpace = svalBuilder.evalBinOpNN(state, BO_Sub, *LenValNL,
-                                                   *dstStrLengthNL, sizeTy);
-          if (!isa<NonLoc>(FreeSpace))
-            break;
-
-          FreeSpace =
-              svalBuilder.evalBinOp(state, BO_Sub, FreeSpace,
-                                    svalBuilder.makeIntVal(1, sizeTy), sizeTy);
-          std::optional<NonLoc> FreeSpaceNL = FreeSpace.getAs<NonLoc>();
-          if (!FreeSpaceNL)
-            break;
-
-          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*FreeSpaceNL);
-          break;
-        }
-        }
       }
     }
 

--- a/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
@@ -2235,14 +2235,14 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
     bool CouldAccessOutOfBound = true;
     if (IsBounded && amountCopied.isUnknown()) {
       // Get the max number of characters to copy.
-      SizeArgExpr lenExpr = {{Call.getArgExpr(2), 2}};
-      SVal lenVal = state->getSVal(lenExpr.Expression, LCtx);
+      const Expr *LenExpr = Call.getArgExpr(2);
+      SVal LenVal = state->getSVal(LenExpr, LCtx);
 
       // Protect against misdeclared strncpy().
-      lenVal =
-          svalBuilder.evalCast(lenVal, sizeTy, lenExpr.Expression->getType());
+      LenVal =
+          svalBuilder.evalCast(LenVal, sizeTy, LenExpr->getType());
 
-      std::optional<NonLoc> lenValNL = lenVal.getAs<NonLoc>();
+      std::optional<NonLoc> LenValNL = LenVal.getAs<NonLoc>();
 
       auto CouldAccessOutOfBoundForSVal = [&](NonLoc Val) -> bool {
         return !isFirstBufInBound(C, state, C.getSVal(Dst.Expression),
@@ -2254,30 +2254,30 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
         CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*strLengthNL);
       }
 
-      if (CouldAccessOutOfBound && lenValNL) {
+      if (CouldAccessOutOfBound && LenValNL) {
         switch (appendK) {
         case ConcatFnKind::none:
         case ConcatFnKind::strcat: {
-          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*lenValNL);
+          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*LenValNL);
           break;
         }
         case ConcatFnKind::strlcat: {
           if (!dstStrLengthNL)
             break;
 
-          SVal freeSpace = svalBuilder.evalBinOpNN(state, BO_Sub, *lenValNL,
+          SVal FreeSpace = svalBuilder.evalBinOpNN(state, BO_Sub, *LenValNL,
                                                    *dstStrLengthNL, sizeTy);
-          if (!isa<NonLoc>(freeSpace))
+          if (!isa<NonLoc>(FreeSpace))
             break;
 
-          freeSpace =
-              svalBuilder.evalBinOp(state, BO_Sub, freeSpace,
+          FreeSpace =
+              svalBuilder.evalBinOp(state, BO_Sub, FreeSpace,
                                     svalBuilder.makeIntVal(1, sizeTy), sizeTy);
-          std::optional<NonLoc> freeSpaceNL = freeSpace.getAs<NonLoc>();
-          if (!freeSpaceNL)
+          std::optional<NonLoc> FreeSpaceNL = FreeSpace.getAs<NonLoc>();
+          if (!FreeSpaceNL)
             break;
 
-          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*freeSpaceNL);
+          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*FreeSpaceNL);
           break;
         }
         }

--- a/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
@@ -2234,15 +2234,16 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
     // situation.
     bool CouldAccessOutOfBound = true;
     if (IsBounded && amountCopied.isUnknown()) {
-      auto CouldAccessOutOfBoundForSVal = [&](NonLoc Val) -> bool {
+      auto CouldAccessOutOfBoundForSVal =
+          [&](std::optional<NonLoc> Val) -> bool {
+        if (!Val)
+          return true;
         return !isFirstBufInBound(C, state, C.getSVal(Dst.Expression),
-                                  Dst.Expression->getType(), Val,
+                                  Dst.Expression->getType(), *Val,
                                   C.getASTContext().getSizeType());
       };
 
-      if (strLengthNL) {
-        CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*strLengthNL);
-      }
+      CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(strLengthNL);
 
       if (CouldAccessOutOfBound) {
         // Get the max number of characters to copy.
@@ -2255,8 +2256,8 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
         // Because analyzer doesn't handle expressions like `size -
         // dstLen - 1` very well, we roughly use `size` for
         // ConcatFnKind::strlcat here, same with other concat kinds.
-        if (std::optional<NonLoc> LenValNL = LenVal.getAs<NonLoc>())
-          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*LenValNL);
+        CouldAccessOutOfBound =
+            CouldAccessOutOfBoundForSVal(LenVal.getAs<NonLoc>());
       }
     }
 

--- a/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
@@ -272,8 +272,7 @@ public:
   static ProgramStateRef
   invalidateDestinationBufferBySize(CheckerContext &C, ProgramStateRef S,
                                     const Expr *BufE, ConstCFGElementRef Elem,
-                                    SVal BufV, SVal SizeV, QualType SizeTy,
-                                    bool CouldAccessOutOfBound = true);
+                                    SVal BufV, SVal SizeV, QualType SizeTy);
 
   /// Operation never overflows, do not invalidate the super region.
   static ProgramStateRef invalidateDestinationBufferNeverOverflows(
@@ -1212,17 +1211,14 @@ bool CStringChecker::isFirstBufInBound(CheckerContext &C, ProgramStateRef State,
 
 ProgramStateRef CStringChecker::invalidateDestinationBufferBySize(
     CheckerContext &C, ProgramStateRef S, const Expr *BufE,
-    ConstCFGElementRef Elem, SVal BufV, SVal SizeV, QualType SizeTy,
-    bool CouldAccessOutOfBound) {
+    ConstCFGElementRef Elem, SVal BufV, SVal SizeV, QualType SizeTy) {
   auto InvalidationTraitOperations =
-      [&C, S, BufTy = BufE->getType(), BufV, SizeV, SizeTy,
-       CouldAccessOutOfBound](RegionAndSymbolInvalidationTraits &ITraits,
-                              const MemRegion *R) {
+      [&C, S, BufTy = BufE->getType(), BufV, SizeV,
+       SizeTy](RegionAndSymbolInvalidationTraits &ITraits, const MemRegion *R) {
         // If destination buffer is a field region and access is in bound, do
         // not invalidate its super region.
         if (MemRegion::FieldRegionKind == R->getKind() &&
-            (!CouldAccessOutOfBound ||
-             isFirstBufInBound(C, S, BufV, BufTy, SizeV, SizeTy))) {
+            isFirstBufInBound(C, S, BufV, BufTy, SizeV, SizeTy)) {
           ITraits.setTrait(
               R,
               RegionAndSymbolInvalidationTraits::TK_DoNotInvalidateSuperRegion);
@@ -2295,9 +2291,13 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
     // can use LazyCompoundVals to copy the source values into the destination.
     // This would probably remove any existing bindings past the end of the
     // string, but that's still an improvement over blank invalidation.
-    state = invalidateDestinationBufferBySize(
-        C, state, Dst.Expression, Call.getCFGElementRef(), *dstRegVal,
-        amountCopied, C.getASTContext().getSizeType(), CouldAccessOutOfBound);
+    if (CouldAccessOutOfBound)
+      state = invalidateDestinationBufferBySize(
+          C, state, Dst.Expression, Call.getCFGElementRef(), *dstRegVal,
+          amountCopied, C.getASTContext().getSizeType());
+    else
+      state = invalidateDestinationBufferNeverOverflows(
+          C, state, Call.getCFGElementRef(), *dstRegVal);
 
     // Invalidate the source (const-invalidation without const-pointer-escaping
     // the address of the top-level region).

--- a/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
@@ -272,7 +272,8 @@ public:
   static ProgramStateRef
   invalidateDestinationBufferBySize(CheckerContext &C, ProgramStateRef S,
                                     const Expr *BufE, ConstCFGElementRef Elem,
-                                    SVal BufV, SVal SizeV, QualType SizeTy);
+                                    SVal BufV, SVal SizeV, QualType SizeTy,
+                                    bool CouldAccessOutOfBound = true);
 
   /// Operation never overflows, do not invalidate the super region.
   static ProgramStateRef invalidateDestinationBufferNeverOverflows(
@@ -1211,14 +1212,17 @@ bool CStringChecker::isFirstBufInBound(CheckerContext &C, ProgramStateRef State,
 
 ProgramStateRef CStringChecker::invalidateDestinationBufferBySize(
     CheckerContext &C, ProgramStateRef S, const Expr *BufE,
-    ConstCFGElementRef Elem, SVal BufV, SVal SizeV, QualType SizeTy) {
+    ConstCFGElementRef Elem, SVal BufV, SVal SizeV, QualType SizeTy,
+    bool CouldAccessOutOfBound) {
   auto InvalidationTraitOperations =
-      [&C, S, BufTy = BufE->getType(), BufV, SizeV,
-       SizeTy](RegionAndSymbolInvalidationTraits &ITraits, const MemRegion *R) {
+      [&C, S, BufTy = BufE->getType(), BufV, SizeV, SizeTy,
+       CouldAccessOutOfBound](RegionAndSymbolInvalidationTraits &ITraits,
+                              const MemRegion *R) {
         // If destination buffer is a field region and access is in bound, do
         // not invalidate its super region.
         if (MemRegion::FieldRegionKind == R->getKind() &&
-            isFirstBufInBound(C, S, BufV, BufTy, SizeV, SizeTy)) {
+            (!CouldAccessOutOfBound ||
+             isFirstBufInBound(C, S, BufV, BufTy, SizeV, SizeTy))) {
           ITraits.setTrait(
               R,
               RegionAndSymbolInvalidationTraits::TK_DoNotInvalidateSuperRegion);
@@ -2223,6 +2227,67 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
         Result = lastElement;
     }
 
+    // For bounded method, amountCopied take the minimum of two values,
+    // for ConcatFnKind::strlcat:
+    // amountCopied = min (size - dstLen - 1 , srcLen)
+    // for others:
+    // amountCopied = min (srcLen, size)
+    // So even if we don't know about amountCopied, as long as one of them will
+    // not cause an out-of-bound access, the whole function's operation will not
+    // too, that will avoid invalidating the superRegion of data member in that
+    // situation.
+    bool CouldAccessOutOfBound = true;
+    if (IsBounded && amountCopied.isUnknown()) {
+      // Get the max number of characters to copy.
+      SizeArgExpr lenExpr = {{Call.getArgExpr(2), 2}};
+      SVal lenVal = state->getSVal(lenExpr.Expression, LCtx);
+
+      // Protect against misdeclared strncpy().
+      lenVal =
+          svalBuilder.evalCast(lenVal, sizeTy, lenExpr.Expression->getType());
+
+      std::optional<NonLoc> lenValNL = lenVal.getAs<NonLoc>();
+
+      auto CouldAccessOutOfBoundForSVal = [&](NonLoc Val) -> bool {
+        return !isFirstBufInBound(C, state, C.getSVal(Dst.Expression),
+                                  Dst.Expression->getType(), Val,
+                                  C.getASTContext().getSizeType());
+      };
+
+      if (strLengthNL) {
+        CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*strLengthNL);
+      }
+
+      if (CouldAccessOutOfBound && lenValNL) {
+        switch (appendK) {
+        case ConcatFnKind::none:
+        case ConcatFnKind::strcat: {
+          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*lenValNL);
+          break;
+        }
+        case ConcatFnKind::strlcat: {
+          if (!dstStrLengthNL)
+            break;
+
+          SVal freeSpace = svalBuilder.evalBinOpNN(state, BO_Sub, *lenValNL,
+                                                   *dstStrLengthNL, sizeTy);
+          if (!isa<NonLoc>(freeSpace))
+            break;
+
+          freeSpace =
+              svalBuilder.evalBinOp(state, BO_Sub, freeSpace,
+                                    svalBuilder.makeIntVal(1, sizeTy), sizeTy);
+          std::optional<NonLoc> freeSpaceNL = freeSpace.getAs<NonLoc>();
+          if (!freeSpaceNL)
+            break;
+
+          CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*freeSpaceNL);
+          break;
+        }
+        }
+      }
+    }
+
     // Invalidate the destination (regular invalidation without pointer-escaping
     // the address of the top-level region). This must happen before we set the
     // C string length because invalidation will clear the length.
@@ -2232,7 +2297,7 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
     // string, but that's still an improvement over blank invalidation.
     state = invalidateDestinationBufferBySize(
         C, state, Dst.Expression, Call.getCFGElementRef(), *dstRegVal,
-        amountCopied, C.getASTContext().getSizeType());
+        amountCopied, C.getASTContext().getSizeType(), CouldAccessOutOfBound);
 
     // Invalidate the source (const-invalidation without const-pointer-escaping
     // the address of the top-level region).

--- a/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp
@@ -2252,12 +2252,10 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
         // Protect against misdeclared strncpy().
         LenVal = svalBuilder.evalCast(LenVal, sizeTy, LenExpr->getType());
 
-        std::optional<NonLoc> LenValNL = LenVal.getAs<NonLoc>();
-
         // Because analyzer doesn't handle expressions like `size -
         // dstLen - 1` very well, we roughly use `size` for
         // ConcatFnKind::strlcat here, same with other concat kinds.
-        if (LenValNL)
+        if (std::optional<NonLoc> LenValNL = LenVal.getAs<NonLoc>())
           CouldAccessOutOfBound = CouldAccessOutOfBoundForSVal(*LenValNL);
       }
     }

--- a/clang/test/Analysis/cstring-should-not-invalidate.cpp
+++ b/clang/test/Analysis/cstring-should-not-invalidate.cpp
@@ -1,4 +1,4 @@
-// RUN: %clang_analyze_cc1 -analyzer-checker=core -verify %s
+// RUN: %clang_analyze_cc1 -analyzer-checker=core,unix.cstring,alpha.unix.cstring -verify %s
 
 // expected-no-diagnostics
 
@@ -63,7 +63,7 @@ struct strncatTestClass {
 
   void KnownLen(char *src) {
     m_ptr = new int;
-    strncat(m_buff, src, sizeof(m_buff)); // known len but unknown src size
+    strncat(m_buff, src, sizeof(m_buff) - 1); // known len but unknown src size
     delete m_ptr;                         // no warning
   }
 

--- a/clang/test/Analysis/cstring-should-not-invalidate.cpp
+++ b/clang/test/Analysis/cstring-should-not-invalidate.cpp
@@ -1,13 +1,18 @@
-// RUN: %clang_analyze_cc1 -analyzer-checker=core,unix.cstring,alpha.unix.cstring -verify %s
-
-// expected-no-diagnostics
+// RUN: %clang_analyze_cc1 -analyzer-checker=core,unix.cstring,alpha.unix.cstring,debug.ExprInspection -verify %s
 
 typedef decltype(sizeof(int)) size_t;
+void clang_analyzer_eval(bool);
 
 char *strncpy(char *dest, const char *src, size_t x);
 
+constexpr int initB = 100;
+struct Base {
+  int b;
+  Base(): b(initB) {}
+};
+
 // issue 143807
-struct strncpyTestClass {
+struct strncpyTestClass: public Base {
   int *m_ptr;
   char m_buff[1000];
 
@@ -28,11 +33,12 @@ void strncpyTest(char *src, size_t n) {
   strncpyTestClass rep;
   rep.KnownLen(src);
   rep.KnownSrcLen(n);
+  clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
 }
 
 size_t strlcpy(char *dest, const char *src, size_t size);
 
-struct strlcpyTestClass {
+struct strlcpyTestClass: public Base {
   int *m_ptr;
   char m_buff[1000];
 
@@ -53,11 +59,12 @@ void strlcpyTest(char *src, size_t n) {
   strlcpyTestClass rep;
   rep.KnownLen(src);
   rep.KnownSrcLen(n);
+  clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
 }
 
 char *strncat(char *s1, const char *s2, size_t n);
 
-struct strncatTestClass {
+struct strncatTestClass: public Base {
   int *m_ptr;
   char m_buff[1000];
 
@@ -78,11 +85,12 @@ void strncatTest(char *src, size_t n) {
   strncatTestClass rep;
   rep.KnownLen(src);
   rep.KnownSrcLen(n);
+  clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
 }
 
 size_t strlcat(char *dst, const char *src, size_t size);
 
-struct strlcatTestClass {
+struct strlcatTestClass: public Base {
   int *m_ptr;
   char m_buff[1000];
 
@@ -103,4 +111,5 @@ void strlcatTest(char *src, size_t n) {
   strlcatTestClass rep;
   rep.KnownLen(src);
   rep.KnownSrcLen(n);
+  clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
 }

--- a/clang/test/Analysis/cstring-should-not-invalidate.cpp
+++ b/clang/test/Analysis/cstring-should-not-invalidate.cpp
@@ -88,6 +88,24 @@ void strncatTest(char *src, size_t n) {
   clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
 }
 
+struct strncatReportOutOfBoundTestClass: public Base {
+  int *m_ptr;
+  char m_buff[1000];
+
+  void KnownLen(char *src) {
+    m_ptr = new int;
+    // expected-warning@+1{{String concatenation function overflows the destination buffer}}
+    strncat(m_buff, src, sizeof(m_buff)); // known len but unknown src size
+    delete m_ptr;                         // no warning
+  }
+};
+
+void strncatReportOutOfBoundTest(char *src, size_t n) {
+  strncatReportOutOfBoundTestClass rep;
+  rep.KnownLen(src);
+  clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
+}
+
 size_t strlcat(char *dst, const char *src, size_t size);
 
 struct strlcatTestClass: public Base {

--- a/clang/test/Analysis/cstring-should-not-invalidate.cpp
+++ b/clang/test/Analysis/cstring-should-not-invalidate.cpp
@@ -88,7 +88,7 @@ void strncatTest(char *src, size_t n) {
   clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
 }
 
-struct strncatReportOutOfBoundTestClass: public Base {
+struct strncatReportOutOfBoundTestClass {
   int *m_ptr;
   char m_buff[1000];
 
@@ -103,7 +103,6 @@ struct strncatReportOutOfBoundTestClass: public Base {
 void strncatReportOutOfBoundTest(char *src, size_t n) {
   strncatReportOutOfBoundTestClass rep;
   rep.KnownLen(src);
-  clang_analyzer_eval(rep.b == initB); // expected-warning{{TRUE}}
 }
 
 size_t strlcat(char *dst, const char *src, size_t size);

--- a/clang/test/Analysis/cstring-should-not-invalidate.cpp
+++ b/clang/test/Analysis/cstring-should-not-invalidate.cpp
@@ -1,0 +1,107 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=core,debug.ExprInspection
+// -analyzer-config c++-inlining=constructors -verify %s
+
+// expected-no-diagnostics
+
+typedef unsigned int size_t;
+
+char *strncpy(char *dest, const char *src, size_t x);
+
+// issue 143807
+struct strncpyTestClass {
+  int *m_ptr;
+  char m_buff[1000];
+
+  void KnownLen(char *src) {
+    m_ptr = new int;
+    strncpy(m_buff, src, sizeof(m_buff)); // known len but unknown src size
+    delete m_ptr;                         // no warning
+  }
+
+  void KnownSrcLen(size_t n) {
+    m_ptr = new int;
+    strncpy(m_buff, "xyz", n); // known src size but unknown len
+    delete m_ptr;              // no warning
+  }
+};
+
+void strncpyTest(char *src, size_t n) {
+  strncpyTestClass rep;
+  rep.KnownLen(src);
+  rep.KnownSrcLen(n);
+}
+
+size_t strlcpy(char *dest, const char *src, size_t size);
+
+struct strlcpyTestClass {
+  int *m_ptr;
+  char m_buff[1000];
+
+  void KnownLen(char *src) {
+    m_ptr = new int;
+    strlcpy(m_buff, src, sizeof(m_buff)); // known len but unknown src size
+    delete m_ptr;                         // no warning
+  }
+
+  void KnownSrcLen(size_t n) {
+    m_ptr = new int;
+    strlcpy(m_buff, "xyz", n); // known src size but unknown len
+    delete m_ptr;              // no warning
+  }
+};
+
+void strlcpyTest(char *src, size_t n) {
+  strlcpyTestClass rep;
+  rep.KnownLen(src);
+  rep.KnownSrcLen(n);
+}
+
+char *strncat(char *s1, const char *s2, size_t n);
+
+struct strncatTestClass {
+  int *m_ptr;
+  char m_buff[1000];
+
+  void KnownLen(char *src) {
+    m_ptr = new int;
+    strncat(m_buff, src, sizeof(m_buff)); // known len but unknown src size
+    delete m_ptr;                         // no warning
+  }
+
+  void KnownSrcLen(size_t n) {
+    m_ptr = new int;
+    strncat(m_buff, "xyz", n); // known src size but unknown len
+    delete m_ptr;              // no warning
+  }
+};
+
+void strncatTest(char *src, size_t n) {
+  strncatTestClass rep;
+  rep.KnownLen(src);
+  rep.KnownSrcLen(n);
+}
+
+size_t strlcat(char *dst, const char *src, size_t size);
+
+struct strlcatTestClass {
+  int *m_ptr;
+  char m_buff[1000];
+
+  void KnownLen(char *src) {
+    m_ptr = new int;
+    strlcat(m_buff, src, sizeof(m_buff)); // known len but unknown src size
+    delete m_ptr;                         // no warning
+  }
+
+  void KnownSrcLen(size_t n) {
+    m_ptr = new int;
+    strlcat(m_buff, "xyz", n); // known src size but unknown len
+    delete m_ptr;              // no warning
+  }
+};
+
+void strlcatTest(char *src, size_t n) {
+  strlcatTestClass rep;
+  rep.KnownLen(src);
+  rep.KnownSrcLen(n);
+}

--- a/clang/test/Analysis/cstring-should-not-invalidate.cpp
+++ b/clang/test/Analysis/cstring-should-not-invalidate.cpp
@@ -1,9 +1,8 @@
-// RUN: %clang_analyze_cc1 -analyzer-checker=core,debug.ExprInspection
-// -analyzer-config c++-inlining=constructors -verify %s
+// RUN: %clang_analyze_cc1 -analyzer-checker=core -verify %s
 
 // expected-no-diagnostics
 
-typedef unsigned int size_t;
+typedef decltype(sizeof(int)) size_t;
 
 char *strncpy(char *dest, const char *src, size_t x);
 

